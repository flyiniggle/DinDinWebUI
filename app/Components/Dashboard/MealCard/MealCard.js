import React from 'react';

import './MealCard.sass';

function MealCard({ meal }) {
    return (
        <div className="mealCard row p-lg-4 p-2 mb-2">
            <h2>{ meal.name }</h2>
            <span>{ meal.taste }</span>
            <span>{ meal.difficulty }</span>
            <h3>{ meal.last_used }</h3>
            <h4>Used { meal.used_count } times</h4>
        </div>
    );
}

export default MealCard;