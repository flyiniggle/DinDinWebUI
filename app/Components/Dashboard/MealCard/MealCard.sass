@import "~Styles/theme.sass"

.mealCard
  border-bottom: 4px solid $secondary
  box-shadow: 7px 7px 5px -4px rgba(0,0,0,0.75)
  transition: 0.5s

  &:hover
    text-decoration: none
    color: initial
    border-bottom: 4px solid $quaternary

  h2
    color: $primary

  h4
    color: $tertiary

  span
    color: $quaternary

  button
    color: $brand-base

    &hover
      background-color: $quaternary