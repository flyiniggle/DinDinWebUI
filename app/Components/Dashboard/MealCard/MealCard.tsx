import { pipe } from 'ramda';
import * as React from 'react';
import { Link } from 'react-router-dom';
import { Maybe } from 'true-myth';
import safeGetProp from 'Business/Lib/safeGetProp';
import Meal from 'Business/Meals/Types/Meal';
import dateString from 'UI/Formatting/dateString';

import './MealCard.sass';
import maybe from 'Business/Lib/maybe';

interface ILastUsedDisplayProps {
    date?: string
}

function renderLastUsed(lastUsedDate: string) {
    return (
        <>
        <h4>Last Used: </h4>
        <span className='lastUsed'>{lastUsedDate}</span>
        </>
    );
};

const displayLastUsed: (date: string) => JSX.Element | null = pipe(
    maybe,
    Maybe.map(dateString.display),
    Maybe.map(renderLastUsed),
    Maybe.getOr(null)
)

interface MealCardProps {
    meal: Meal
    useMeal: (meal: Meal) => void,
    mealIsUpdating: boolean
}

function MealCard(props: MealCardProps) {
    const { meal, useMeal, mealIsUpdating: isMealUpdating } = props;
    const handleUseMeal = function (e: React.MouseEvent<HTMLButtonElement>) {
        e.stopPropagation();
        e.preventDefault();

        useMeal(meal);
    }

    return (
        <Link to={`/meals/${meal.id}`} className="mealCard row p-lg-4 p-2 mb-2">
            <div className="col-4">
                <h2>{meal.name}</h2>
                { displayLastUsed(meal.lastUsed) }
            </div>
            <div className="col-4">
                <div><h4 className="d-inline">Taste: </h4><span className="taste">{meal.taste}</span></div>
                <div><h4 className="d-inline">Difficulty: </h4><span className="difficulty">{meal.difficulty}</span></div>
            </div>
            <div className="col-4">
                <button
                    className="btn btn-primary useMealButton"
                    type="button"
                    onClick={handleUseMeal}>Use it!</button>
                <h4 className="usedCount">Used {meal.usedCount} {(meal.usedCount === 1) ? 'time' : 'times'}</h4>
            </div>
        </Link>
    );
}


export default MealCard;