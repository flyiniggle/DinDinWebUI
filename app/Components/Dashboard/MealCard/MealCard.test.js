import React from 'react';
import { shallow } from 'enzyme';

import { MealCardBase as MealCard } from './MealCard';

describe('#Components #Dashboard #MealCard', function() {
    const meal = {
        id: 1,
        name: 'Delicious Meal',
        owner: '1',
        taste: 5,
        difficulty: 3,
        lastUsed: '2018-05-12',
        usedCount: 3,
        notes: 'this is a really good meal'
    };

    it('should render.', function() {
        expect(shallow(<MealCard meal={ meal } />)).toMatchSnapshot();
    });

    it('should show the meal name.', function() {
        const wrapper = shallow(<MealCard meal={ meal } />);

        expect(/Delicious Meal/.test(wrapper.text()));
    });

    it('should show the date the meal was last used.', function() {
        const wrapper = shallow(<MealCard meal={ meal } />);

        expect(wrapper.find('.lastUsed').text()).toEqual('May 12, 2018');
    });

    it('should show the meal taste rating.', function() {
        const wrapper = shallow(<MealCard meal={ meal } />);

        expect(wrapper.find('.taste').text()).toEqual('5');
    });

    it('should show the difficulty rating.', function() {
        const wrapper = shallow(<MealCard meal={ meal } />);

        expect(wrapper.find('.difficulty').text()).toEqual('3');
    });

    it('should show the numbe of times the meal was used.', function() {
        expect(shallow(<MealCard meal={ meal } />).find('.usedCount').text()).toEqual('Used 3 times');

        meal.usedCount = 1;
        expect(shallow(<MealCard meal={ meal } />).find('.usedCount').text()).toEqual('Used 1 time');

        meal.usedCount = 0;
        expect(shallow(<MealCard meal={ meal } />).find('.usedCount').text()).toEqual('Used 0 times');
    });

    it('should call the useMeal handler when clicking the appropriate button.', function() {
        const useMealSpy = jest.fn();
        const wrapper = shallow(<MealCard meal={ meal } useMeal={ useMealSpy } />);

        wrapper.find('.useMealButton').simulate('click', { stopPropagation: () => undefined, preventDefault: () => undefined });

        expect(useMealSpy).toHaveBeenCalledWith(meal);
    });
});