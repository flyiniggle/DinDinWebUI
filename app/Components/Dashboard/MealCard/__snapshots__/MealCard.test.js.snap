// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Components #Dashboard #MealCard should render. 1`] = `
<Link
  className="mealCard row p-lg-4 p-2 mb-2"
  replace={false}
  to="/meals/1"
>
  <div
    className="col-4"
  >
    <h2>
      Delicious Meal
    </h2>
    <React.Fragment>
      <h4>
        Last Used: 
      </h4>
      <span
        className="lastUsed"
      >
        May 12, 2018
      </span>
    </React.Fragment>
  </div>
  <div
    className="col-4"
  >
    <div>
      <h4
        className="d-inline"
      >
        Taste: 
      </h4>
      <span
        className="taste"
      >
        5
      </span>
    </div>
    <div>
      <h4
        className="d-inline"
      >
        Difficulty: 
      </h4>
      <span
        className="difficulty"
      >
        3
      </span>
    </div>
  </div>
  <div
    className="col-4"
  >
    <button
      className="btn btn-primary useMealButton"
      onClick={[Function]}
      type="button"
    >
      Use it!
    </button>
    <h4
      className="usedCount"
    >
      Used 
      3
       
      times
    </h4>
  </div>
</Link>
`;
